src/lib/core/datetime.h

```diff
@@ -400,6 +415,24 @@ datetime_usec(const struct datetime *date)
 	return datetime_nsec(date) / 1000;
 }
 
+/** Simplified checker for datetime structure validity */
+static inline bool
+datetime_validate(const struct datetime *date)
+{
+	if (unlikely(date->epoch < MIN_EPOCH_SECS_VALUE) ||
+	    unlikely(date->epoch > MAX_EPOCH_SECS_VALUE))
+		return false;
+	if (unlikely(date->nsec < 0) || unlikely(date->nsec > NANOS_PER_SEC))
```

Comment:
IIRC, there were some cases (maybe not present now, but planned), when date->nsec could be `2 000 000 000` due to a leap second.
Should we allow it here? Or did I mix things up?

Reply:
That's very good question: yes, originally we assumed that once we would need to handle some specific leap second then it will be done via nanoseconds part which would be allowed to handle 2e9 nanoseconds.

But during integration of tzcode in previous stages we have discovered that if we would allow leap-seconds in tzcode epoch then we would receive epoch values different to system. (And in POSIX epoch we do not save leap-seconds at all, they are non-existing in the range of posix values). tzcode, if being compiled with precise leap-seconds support, does have routine to convert between their own epoch values and system epoch, with skipped leap-seconds. But, for simplicity sake, we have opted to not behave differently to the surrounding system, and behave as it's done in posix functions.

That's why there is no currently way to receive automatically constructed/parsed date/time value where leap-seconds would be larger than 1e9. (And if that's incorrect we do have a couple of assertions here and there). That's why it may be still ok to accept only upto 1e9, not 2e9.

But in a longer term yes, we need to create `MAX_NANOS_PER_SEC` as 2e9, and apply it here and in parsers.

Reply:
Not sure (yet) it's a blocker for current pulle-request.

Reply:
Ok. What happens if we allow nsec up to `2e9` now and receive a date with `nsec == 2e9`? Will some assertion fire once we try to process it?

If everything will be fine, I suggest to allow `2e9` in nsec right now. So that possible future extension is accepted by current Tarantool versions.

If something will break right now when we try to process `2e9`, then let's disallow it.

Reply:
```diff
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 8d8d96da9..6f41c2f3a 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -32,6 +32,7 @@ struct tnt_tm;
 #define SECS_PER_DAY          86400
 #define DT_EPOCH_1970_OFFSET  719163
 #define NANOS_PER_SEC         1000000000LL
+#define MAX_NANOS_PER_SEC     2000000000LL
 #endif
 
 /** Required size of datetime_to_string string buffer */
@@ -422,7 +423,8 @@ datetime_validate(const struct datetime *date)
 	if (unlikely(date->epoch < MIN_EPOCH_SECS_VALUE) ||
 	    unlikely(date->epoch > MAX_EPOCH_SECS_VALUE))
 		return false;
-	if (unlikely(date->nsec < 0) || unlikely(date->nsec > NANOS_PER_SEC))
+	if (unlikely(date->nsec < 0) ||
+	    unlikely(date->nsec > MAX_NANOS_PER_SEC))
 		return false;
 	if (unlikely(date->tzoffset < MIN_TZOFFSET) ||
 	    unlikely(date->tzoffset > MAX_TZOFFSET))
diff --git a/src/lib/tzcode/strftime.c b/src/lib/tzcode/strftime.c
index b6bfbf1ea..9bb437e24 100644
--- a/src/lib/tzcode/strftime.c
+++ b/src/lib/tzcode/strftime.c
@@ -35,6 +35,7 @@
 
 #include "private.h"
 #include "trivia/util.h"
+#include "datetime.h"
 #include "tzcode.h"
 #include "timezone.h"
 #include "timelocal.h"
@@ -281,7 +282,7 @@ _fmt(char *buf, ssize_t size, const char *format, const struct tnt_tm *t,
 				/* fallthru */
 			case 'f': {
 				int nsec = t->tm_nsec;
-				assert(nsec < 1000000000);
+				assert(nsec < MAX_NANOS_PER_SEC);
 
 				/*
 				 ** no length modifier provided -
```